packages/instrumentation-openai/src/instrumentation.ts

```diff
@@ -0,0 +1,748 @@
+        unit: '{token}',
+        advice: {
+          explicitBucketBoundaries: [
+            1, 4, 16, 64, 256, 1024, 4096, 16384, 65536, 262144, 1048576,
+            4194304, 16777216, 67108864,
+          ],
+        },
+      }
+    );
+  }
+
+  _getPatchedChatCompletionsCreate() {
+    const self = this;
+    return (original: (...args: unknown[]) => any) => {
+      // https://platform.openai.com/docs/api-reference/chat/create
+      return function patchedCreate(this: any, ...args: unknown[]) {
+        if (!self.isEnabled) {
+          return original.apply(this, args);
+        }
+
+        debug('OpenAI.Chat.Completions.create args: %O', args);
+        const params =
+          args[0] as any; /* type ChatCompletionCreateParamsStreaming */
+        const config = self.getConfig();
+        const startNow = performance.now();
+
+        let startInfo;
+        try {
+          startInfo = self._startChatCompletionsSpan(
+            params,
+            config,
+            this?._client?.baseURL
+          );
+        } catch (err) {
+          self._diag.error('unexpected error starting span:', err);
+          return original.apply(this, args);
+        }
+        const { span, ctx, commonAttrs } = startInfo;
+
+        const apiPromise: Promise<any> = context.with(ctx, () =>
+          original.apply(this, args)
+        );
+
+        // Streaming.
+        if (params && params.stream) {
+          // When streaming, `apiPromise` resolves to `import('openai/streaming').Stream`,
+          // an async iterable (i.e. has a `Symbol.asyncIterator` method). We
+          // want to wrap that iteration to gather telemetry. Instead of wrapping
+          // `Symbol.asyncIterator`, which would be nice, we wrap the `iterator`
+          // method because it is used internally by `Stream#tee()`.
+          return apiPromise.then(stream => {
+            self._wrap(stream, 'iterator', origIterator => {
+              return () => {
+                return self._onChatCompletionsStreamIterator(
+                  origIterator(),
+                  span,
+                  startNow,
+                  config,
+                  commonAttrs,
+                  ctx
+                );
+              };
+            });
+            return stream;
+          });
+        }
+
+        // Non-streaming.
+        apiPromise
+          .then(result => {
+            self._onChatCompletionsCreateResult(
+              span,
+              startNow,
+              commonAttrs,
+              result,
+              config,
+              ctx
+            );
+          })
+          .catch(
+            self._createAPIPromiseRejectionHandler(startNow, span, commonAttrs)
+          );
+
+        return apiPromise;
+      };
+    };
+  }
+
+  /**
+   * Start a span for this chat-completion API call. This also emits log events
+   * as appropriate for the request params.
+   *
+   * @param {import('openai').OpenAI.ChatCompletionCreateParams} params
```

Comment:
I've found it has its drawbacks:
- the import leaks to the exported types and this may give versioning issues in the future
- `typescript@4.4.4` has compilation errors (which go away with latest 4.x)

perhaps having our own type with the properties we are interested in could be a way be more TS friendly

Reply:
My understanding is that if we want strong typing of the OpenAI params and responses being handled in the instrumentation that we'd have to define our own types (in src/internal-types.ts or where ever). We don't want a runtime dep on openai, and I'm not sure we even want a compile-type dep on the `openai` package.

These would be (mostly) a copy of OpenAI types --- frozen to a particular `openai` version.

I wasn't sure this was worth the added effort (and then maintenance?) of extracting those types from a particular version of `openai`. For example, the `ChatCompletionCreateParams` type is 100s of lines, involving some deprecated old edge cases, streaming and non-streaming slight differences.

https://gist.github.com/trentm/e46a8862c1c4221675320aad2b13636e#file-completions-d-ts-L716

Our version of the type "with the properties we are interested in" would be smaller than that, yes.

I guess I don't know where the line is here: In authoring `type OpenAIChatCompletionCreateParams` in our internal-types.ts, am I improving the instrumentation code, or am I taking time away from adding tests and other maintenance to avoid having one more `any` in the TypeScript code? :)

Reply:
deciding where is "the line" is up to the instrumentation author so I'm okay with a couple of anys